### PR TITLE
TINY-10605: Added EditorOptions debug API

### PR DESCRIPTION
**Pull request by danoaky-tiny**
Related Ticket: TINY-10605

Description of Changes:
- Added a `debug` method to `EditorOptions`. It's functionally the same as what's described in the investigation ticket [TINY-10862](https://ephocks.atlassian.net/browse/TINY-10862).
- Added an extra optional parameter to `createOptions` to pass in the initial non-normalised options.
- `normalizeOptions` now doesn't modify its `options` param as a side effect.

Example:
```console
> tinymce.activeEditor.options.debug()
{
    "selector": "textarea.tinymce",
    "plugins": "table lists image accordion code",
    "toolbar": "table | numlist bullist | image | accordion | code",
    "menu": {
        "insert": {
            "title": "Insert",
            "items": "table | image | accordion"
        }
    },
    "details_initial_state": "inherited",
    "details_serialize_state": "inherited",
    "height": 500,
    "paste_as_text": true,
    "inline": null,
    "setup": "[object Function]",
    "encoding": "[object Undefined]",
    "protect": "[object RegExp]",
    "target": "[object HTMLDivElement]"
}

```

Pre-checks:
* [-] Changelog entry added
* [x] Tests have been added (if applicable)
* [-] Branch prefixed with `feature/`, `hotfix/` or `spike/`

Review:
* [-] Milestone set
* [x] Docs ticket created (if applicable)

[TINY-10862]: https://ephocks.atlassian.net/browse/TINY-10862?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by danoaky-tiny** [comment]
Unsure if this needed tests. I could wrap `console.log` & `console.error` and test against that. Let me know.

File: modules/tinymce/src/core/main/ts/api/EditorOptions.ts [hunk 1]
```diff
@@ -147,6 +147,14 @@ export interface Options {
    * @return {Boolean} True if the option has a value set, otherwise false.
    */
   isSet: (name: string) => boolean;
+
+  /**
+   * Logs the initial raw editor options to the console.
+   *
+   * @method debug
+   * @returns {void}
```
Comment (spocke) [inline modules/tinymce/src/core/main/ts/api/EditorOptions.ts:155]: I think we just omit the return if it's void.

File: .changes/unreleased/tinymce-TINY-10605-2024-05-03.yaml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+project: tinymce
+kind: Added
+body: Added `options.debug` API
```
Comment (spocke) [inline .changes/unreleased/tinymce-TINY-10605-2024-05-03.yaml:3]:
```suggestion
body: Added `options.debug` API.
```
Or maybe something like:
```suggestion
body: Added `options.debug` API that logs the initial raw editor options to console.
```
I guess we flush it out in the release notes anyway.

**Review by spocke — APPROVED** [review]
I think we should have tests on this I think we do some console.log monkey patching for some other things seems like this is something want to do here as well since we might need to tweak the output.

**Comment by TheSpyder** [comment]
I agree, we will use this in support so we need to make sure it continues to include everything we expect it to.

**Comment by danoaky-tiny** [comment]
Added tests @TheSpyder @spocke

File: modules/tinymce/src/core/test/ts/browser/EditorOptionsDebugTest.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+import Editor from 'tinymce/core/api/Editor';
+
+describe('browser.tinymce.core.EditorOptionsDebugTest', () => {
+  const createHook = (options: Record<string, unknown>) =>
+    TinyHooks.bddSetupLight<Editor>({
+      ...options,
+      base_url: '/project/tinymce/js/tinymce',
+    }, []);
+
+  const METHODS = [ 'log', 'error' ] as const;
```
Comment (spocke) [inline modules/tinymce/src/core/test/ts/browser/EditorOptionsDebugTest.ts:16]: Seems overly complicated. Should be possible to test this with just:
```
const log: any[] = [];
const orgConsoleLog = console.log;
console.log = (...args) => log.push(args);
editor.options.dump();
console.log = orgConsoleLog;
assert.deepEqual(log, [['foo']])
```

File: modules/tinymce/src/core/test/ts/browser/EditorOptionsDebugTest.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+  context('With empty options', () => {
+    const hook = createHook({});
+    it('TINY-10605: should log empty object', () => {
+      hook.editor().options.debug();
+      assertLastLog({});
+    });
+  });
+
+  context('With function', () => {
```
Comment (ltrouton) [inline modules/tinymce/src/core/test/ts/browser/EditorOptionsDebugTest.ts:66]: I know we can't be exhaustive but could we test setting a random option to `undefined`, `null`, Regexp, and a DOM node. That should cover the majority of the possible option types I would think.

**Review by ltrouton — APPROVED** [review]

File: modules/tinymce/src/core/test/ts/browser/EditorOptionsDebugTest.ts (hunk 4 above)
Comment (danoaky-tiny) [reply modules/tinymce/src/core/test/ts/browser/EditorOptionsDebugTest.ts:38]: Added test coverage for undefined, null, RegExp and a div element.

File: modules/tinymce/src/core/test/ts/browser/EditorOptionsDebugTest.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+      ...options,
+      base_url: '/project/tinymce/js/tinymce',
+    }, []);
+  const ignoredKeys = new Set([ 'promotion', 'toolbar', 'menubar', 'statusbar', 'base_url', 'selector', 'setup' ]);
+  const assertDebugLog = (editor: Editor, expectedLog: unknown) => {
+    const logs: unknown[] = [];
+    const originalLog = console.log;
+    console.log = (arg: unknown) => logs.push(arg);
+    editor.options.debug();
+    console.log = originalLog;
```
Comment (TheSpyder) [inline modules/tinymce/src/core/test/ts/browser/EditorOptionsDebugTest.ts:22]: Please put some sort of try/finally block here, so if `.debug()` throws an exception it doesn't break `console.log` for the entire page

**Review by TheSpyder — APPROVED** [review]

File: modules/tinymce/src/core/test/ts/browser/EditorOptionsDebugTest.ts (hunk 5 above)
Comment (spocke) [reply modules/tinymce/src/core/test/ts/browser/EditorOptionsDebugTest.ts:22]: We have a few spots of these types of collections maybe we should add a transact type collect function in agar. Then we would be able to add that in a central place.

**Comment by TheSpyder** [comment]
I must've submitted my review seconds before it was merged, the timestamp still says my comment was before it merged even though it's below the merge in the timeline 😂

File: modules/tinymce/src/core/test/ts/browser/EditorOptionsDebugTest.ts (hunk 5 above)
Comment (TheSpyder) [reply modules/tinymce/src/core/test/ts/browser/EditorOptionsDebugTest.ts:22]: I was expecting an `after` hook to restore the global values, which is exactly what would be good to do with that sort of collect function

Comment (spocke) [reply modules/tinymce/src/core/test/ts/browser/EditorOptionsDebugTest.ts:22]: That would be really nice something like `const { getMessages, reset } = ConsoleHooks.logs({ levels: [ 'error', 'info' ] })` or something.

Comment (spocke) [reply modules/tinymce/src/core/test/ts/browser/EditorOptionsDebugTest.ts:22]: Maybe a training ticket we could make out of that seems simple enough.

Comment (danoaky-tiny) [reply modules/tinymce/src/core/test/ts/browser/EditorOptionsDebugTest.ts:22]: @TheSpyder  The debug method itself has a try catch encompassing the whole function, so it should be fine as for how it's implemented right now.
I had an after hook in a previous commit, but simplified it later.
Something in agar sounds good to me.

**Pull request merged**